File: src/main/java/fr/uiytt/ageofempire/base/Building.java
package fr.uiytt.ageofempire.base;

import fr.uiytt.ageofempire.ConfigManager;
import org.bukkit.Bukkit;
import org.bukkit.ChatColor;
import org.bukkit.Location;
import org.bukkit.attribute.Attribute;
import org.bukkit.entity.EntityType;
import org.bukkit.entity.Player;
import org.bukkit.entity.Villager;

import java.util.UUID;

public class Building {

    private final BuildingType buildingType;
    private final TeamBase teamBase;

    private boolean inConstruction = false;
    private boolean constructed = false;
    private long lastWarning = System.currentTimeMillis();
    private Location plotLocation = new Location(ConfigManager.getWorld(), 0, 0, 0);
    private double health;

    /**
     * An object storing all the information related to a building constructed in game.
     * @param buildingType The type of building
     * @param teamBase The base in which the building is constructed
     */
    public Building(BuildingType buildingType, TeamBase teamBase) {
        this.buildingType = buildingType;
        this.teamBase = teamBase;
        this.health = buildingType.getHealth();
    }

    public boolean isAvailable() {
        return !inConstruction && !constructed;
    }

    public void summonBuildingVillager(Location villagerLocation) {
        Villager villager = (Villager) villagerLocation.getWorld().spawnEntity(villagerLocation, EntityType.VILLAGER);
        villager.setHealth(20);
        villager.setCustomName(teamBase.getGameTeam().getColor().getChatColor() + buildingType.getDisplayName() + ChatColor.GRAY + " - " + ChatColor.GREEN + villager.getHealth());
        villager.setAI(false);
        villager.getAttribute(Attribute.GENERIC_KNOCKBACK_RESISTANCE).setBaseValue(100);

        teamBase.getGameTeam().getPlayersUUIDs().forEach(playerUUID -> {
            Player player = Bukkit.getPlayer(playerUUID);
            if(player != null) {
                player.sendMessage("Bâtiment " + buildingType.getDisplayName() + " construit.");
            }
        });
    }

    public void sendWarning() {
        if(System.currentTimeMillis() <= lastWarning + 10000) return;
        lastWarning = System.currentTimeMillis();
        for(UUID playerUUID : teamBase.getGameTeam().getPlayersUUIDs()) {
            Player player = Bukkit.getPlayer(playerUUID);
            if(player != null) player.sendMessage(ChatColor.RED + "Votre " + buildingType.getDisplayName() + " est attaqué !!");
        }
    }

    public void explodeBuilding(Location villagerLoc) {
        Plot plot = teamBase.getPlots().get(plotLocation);
        plot.setDestroyed(true);
        plot.setConstructed(false);

        villagerLoc.getWorld().spawnEntity(villagerLoc, EntityType.PRIMED_TNT);
    }
    
    public void setInConstruction(boolean inConstruction) {
        this.inConstruction = inConstruction;
    }
    public void setConstructed(boolean constructed) {
        this.constructed = constructed;
    }
    public Location getPlotLocation() {
        return plotLocation;
    }
    public void setPlotLocation(Location plotLocation) {
        this.plotLocation = plotLocation;
    }
    public double getHealth() {return health;}
    public void setHealth(double health) {this.health = health;}
}

File: src/main/kotlin/fr/uiytt/ageofempire/base/Building.kt
package fr.uiytt.ageofempire.base

import fr.uiytt.ageofempire.AgeOfEmpire
import fr.uiytt.ageofempire.game.getGameManager
import fr.uiytt.ageofempire.game.getPlayerTeam
import org.bukkit.Bukkit
import org.bukkit.ChatColor
import org.bukkit.Location
import org.bukkit.attribute.Attribute
import org.bukkit.entity.EntityType
import org.bukkit.entity.Villager
import java.util.*
import java.util.function.Consumer

/**
 * A building store and manage different type of
 * information for a certain [BuildingType]
 * All [BuildingType] should have a building
 * even if not yet constructed
 * @param buildingType a type of building
 * @param teamBase the base to which the building belong
 */
class Building(val buildingType: BuildingType, private val teamBase: TeamBase) {
    var inConstruction = false
    var isConstructed = false
    var plotLocation = Location(getGameManager().world, 0.0, 0.0, 0.0)
    var health: Double = buildingType.health
    var villager: Villager? = null
    val isAvailable: Boolean
        get() = !inConstruction && !isConstructed
    private var timeOfLastWarning = System.currentTimeMillis()

    fun summonBuildingVillager(villagerLocation: Location, silent: Boolean = false) {
        villager = villagerLocation.world!!.spawnEntity(villagerLocation, EntityType.VILLAGER) as Villager
        villager!!.health = 20.0
        villager!!.customName = teamBase.gameTeam.color.chatColor.toString() + buildingType.displayName + ChatColor.GRAY + " - " + ChatColor.GREEN + health
        villager!!.setAI(false)
        villager!!.getAttribute(Attribute.GENERIC_KNOCKBACK_RESISTANCE)!!.baseValue = 100.0
        if(!silent) {
            teamBase.gameTeam.playersUUIDs.forEach(Consumer { playerUUID: UUID? ->
                val player = Bukkit.getPlayer(playerUUID!!)
                player?.sendMessage("Bâtiment " + buildingType.displayName + " construit.")
            })
        }

        if (buildingType == BuildingType.TEMPLE){
            AgeOfEmpire.gameManager.triggerTemple(teamBase.gameTeam)
        }
    }

    fun sendWarning() {
        if (System.currentTimeMillis() <= timeOfLastWarning + 10000) return
        timeOfLastWarning = System.currentTimeMillis()
        for (playerUUID in teamBase.gameTeam.playersUUIDs) {
            val player = Bukkit.getPlayer(playerUUID)
            player?.sendMessage(ChatColor.RED.toString() + "Votre " + buildingType.displayName + " est attaqué !!")
        }
    }

    fun explodeBuilding() {
        this.isConstructed = false
        val plot = teamBase.plots[plotLocation]!!
        plot.destroyed = true
        plot.constructed = false
        villager!!.location.world!!.spawnEntity(villager!!.location, EntityType.PRIMED_TNT)

        if(buildingType == BuildingType.TEMPLE){
            AgeOfEmpire.gameManager.destroyTemple()
        }
        villager = null
    }

}